src/include/utilities.h

```diff
@@ -106,6 +106,30 @@ FILE *fstream_tee(FILE *orig, FILE *copy);
     pthread_cleanup_pop(0); \
 }
 
+typedef sem_t *pshared_sem_t;
+static inline int pshared_sem_post(sem_t *sem)
+{
+    return sem_post(sem);
+}
+static inline int pshared_sem_wait(sem_t *sem)
+{
+    return sem_wait(sem);
+}
+
+/* macOS doesn't support sem_init(), so use Mach semaphores instead */
+#ifdef __APPLE__
+#include <mach/mach_init.h>
+#include <mach/task.h>
+#include <mach/semaphore.h>
+#undef PAGE_SIZE
+#define PAGE_SIZE 4096
+#define sem_t semaphore_t
+#define sem_init(x,y,z) semaphore_create(mach_task_self(), x, SYNC_POLICY_FIFO, z)
+#define sem_post(x) semaphore_signal(*(x))
+#define sem_wait(x) do {semaphore_wait(*(x)); pthread_testcancel();} while(0)
```

Comment:
I don't understand the trick here.
If semaphore_wait() is not cancelable,
then how would pthread_testcancel()
ever help?

Reply:
Maybe semaphore_wait() returns something
that you need to check?

Reply:
according to `man pthread_testcancel` on mac:

```
   Cancellation Points
     Cancellation points will occur when a thread is executing the following functions: accept(), aio_suspend(), close(), connect(), creat(), fcntl(),
     fsync(), lockf(), msgrcv(), msgsnd(), msync(), nanosleep(), open(), pause(), poll(), pread(), pselect(), pthread_cond_timedwait(),
     pthread_cond_wait(), pthread_join(), pthread_testcancel(), pwrite(), read(), readv(), recv(), recvfrom(), recvmsg(), select(), sem_wait(), send(),
     sendmsg(), sendto(), sigpause(), sigsuspend(), sigwait(), sleep(), system(), tcdrain(), usleep(), wait(), waitpid(), write(), writev().
```
so I interpreted that as with semaphore_wait, unlike with sem_wait there is no cancellation point and in my testing the thread did exit semaphore_wait but then just continued in the loop without exiting, so it entered semaphore_wait again.

But, and I need to check the timing of that... `semaphore_destroy` will cause `semaphore_wait` to exit (see http://web.mit.edu/darwin/src/modules/xnu/osfmk/man/semaphore_destroy.html)
Then the thread calls `pthread_testcancel` and will exit because the thread was cancelled.

Reply:
so yes, if semaphore_wait returns something when the semaphore is destroyed it can be checked.
I'll try and figure it out.

Reply:
It should return something on cancel as
well, after which you might be able to call
pthread_exit().

Reply:
yes indeed it returns KERN_ABORTED on cancel (http://web.mit.edu/darwin/src/modules/xnu/osfmk/man/semaphore_wait.html), so it's sufficient to check for that